content/blog/2019-12-09-comp-cert/index.md

```diff
@@ -0,0 +1,111 @@
+
+<img src="structure.png" style="width: 100%">
+
+In total, CompCert formally defines 8 intermediate languages and 14 passes over them. The 14 passes must be proven to preserve the semantics of the original program. The first few passes simplify the C code by converting all types to either ints or floats (pointers get converted to ints) and explicitly describing memory accesses. The result of these passes is an intermediate representation called Cminor, and below is an example of a translation from Clight to Cminor.
+
+<img src="transf_ex.png" style="width: 100%">
+
+As you can see, function signatures have been made explicit, implicit casts have been made explicit (like the cast from float to int), array accesses have been transformed into exact byte offsets from pointers, and the size of the function's activation record has been made explicit (this is for dealing with the address of operator, which requires function local variables to be mapped to a location on the stack frame of the function).
+
+Next, CompCert performs instruction selection for the specific architecture that is targeting. This is done via instruction tiling which can recognize basic algebraic identities. For example, the instruction selection pass will transform `8 + (x + 1) × 4` into `x × 4 + 12`. These algebraic identities are proven in CompCert in order to assist in the semantic preservation proof. The selected instructions are very similar to available PowerPC instructions. Next, CompCert makes the control flow of the program more explicit via a transformation to an RTL which represents control flow using a CFG. In addition to generating a CFG, the RTL representation transforms variables into pseudo-registers, of which there are an unlimited supply of. The RTL representation is a convenient representation to perform optimizations on, so CompCert runs  several dataflow analyses on the program in order to perform optimizations such as constant propagation, common sub expression elimination, and lazy code motion.
+
+The next transformation pass performed by CompCert maps the pseudo registers to hardware registers or abstract stack locations using a register allocation algorithm. The register allocation algorithm uses the validator method and is discussed in a bit more detail later in the paper. The algorithm implements an approximation of the the coloring algorithm in OCaml which is used in CompCert using the validator method. Further passes linearize the CFG, spill registers on the stack and insert the necessary loads for temporaries. Some simple optimizations like branch tunneling (removing branches to branches) are also performed as part of these passes. Finally, CompCert performs instruction scheduling to increase instruction level parallelism on super scaler processors, such as PowerPC processors, and generates PowerPC assembly code.
```

Comment:
I don't think it's relevant that it's discussed "later in the paper," because when we're reading your blog post, we don't know where in the paper things are. :smiley: Maybe you mean "later in this post," i.e., later in _your_ writing?